NewareNDA/NewareNDAx.py

```diff
@@ -137,48 +139,41 @@ def _data_interpolation(df):
     Some ndax from from BTS Server 8 do not seem to contain a complete dataset.
     This helper function fills in missing times, capacities, and energies.
     """
-    logger.warning("IMPORTANT: This ndax has missing data. The output from "
-                   "NewareNDA contains interpolated data!")
-
+    logger.info(
+        ".ndax is using compact data format, "
+        "reconstructing the data in time, timestamp, capacity, and energy."
+    )
     # Identify the valid data
-    nan_mask = df['Time'].notnull()
-
-    # Group by step and run 'inside' interpolation on Time
-    df['Time'] = df.groupby('Step')['Time'].transform(
-        lambda x: pd.Series.interpolate(x, limit_area='inside'))
-    df['Timestamp'] = df.groupby('Step')['Timestamp'].transform(
-        lambda x: pd.Series.interpolate(x, limit_area='inside'))
+    nan_mask = df['Time'].notnull()  # 1 = valid, 0 = missing
+    nan_groups = nan_mask.cumsum().shift(fill_value=0)  # contiguous nans = same group number
 
-    # Perform extrapolation to generate the remaining missing Time
-    nan_mask2 = df['Time'].notnull()
-    time_inc = df['Time'].diff().ffill().groupby(nan_mask2.cumsum()).cumsum()
-    time = df['Time'].ffill() + time_inc.shift()
-    df['Time'] = df['Time'].where(nan_mask2, time)
+    # Forward fill time differences
+    df['dt'] = df['dt'].ffill()
 
-    # Fill in missing Timestamps
-    time_inc = df['Timestamp'].diff().ffill().groupby(nan_mask2.cumsum()).cumsum()
-    timestamp = df['Timestamp'].ffill() + time_inc.shift()
-    df['Timestamp'] = df['Timestamp'].where(nan_mask2, timestamp)
+    # The time to add is the cumulative sum of dt over the nan groups
+    time_inc = df['dt'].groupby(nan_groups).cumsum() * ~nan_mask
+    df['Time'] = df['Time'].ffill() + time_inc
+    df['Timestamp'] = df['Timestamp'].ffill() + pd.to_timedelta(time_inc, unit='s')
 
     # Integrate to get capacity and fill missing values
-    capacity = df['Time'].diff()*abs(df['Current(mA)'])/3600
-    inc = capacity.groupby(nan_mask.cumsum()).cumsum()
-    chg = df['Charge_Capacity(mAh)'].ffill() + \
-        inc.where(df['Current(mA)'] > 0, 0).shift()
-    dch = df['Discharge_Capacity(mAh)'].ffill() + \
-        inc.where(df['Current(mA)'] < 0, 0).shift()
-    df['Charge_Capacity(mAh)'] = df['Charge_Capacity(mAh)'].where(nan_mask, chg)
-    df['Discharge_Capacity(mAh)'] = df['Discharge_Capacity(mAh)'].where(nan_mask, dch)
+    capacity = df['dt'] * df['Current(mA)'] / 3600
+    capacity_inc = capacity.groupby(nan_groups).cumsum() * ~nan_mask
+    df['Charge_Capacity(mAh)'] = (
+        df['Charge_Capacity(mAh)'].ffill().abs() + capacity_inc.clip(lower=0).abs()
+    )
+    df['Discharge_Capacity(mAh)'] = (
+        df['Discharge_Capacity(mAh)'].ffill().abs() + capacity_inc.clip(upper=0).abs()
+    )
 

```

Comment:
After more testing, these modifications introduce some subtle changes that improve the agreement with BTSDA.